naff/client/auto_shard_client.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import logging
+import time
+from collections import defaultdict
+from typing import (
+    TYPE_CHECKING,
+)
+
+import naff.api.events as events
+from naff.client.const import logger_name, MISSING
+from naff.models import (
+    Guild,
+    to_snowflake,
+)
+from naff.models.naff.listener import Listener
+from naff.client.client import Client
+from naff.api.gateway.state import ConnectionState
+
+if TYPE_CHECKING:
+    from naff.models import Snowflake_Type
+
+__all__ = ("AutoShardedClient",)
+
+from ..api.gateway.gateway import GatewayClient
+
+log = logging.getLogger(logger_name)
+
+
+class AutoShardedClient(Client):
+    def __init__(self, *args, **kwargs) -> None:
+        if "total_shards" not in kwargs:
+            self.auto_sharding = True
+        else:
+            self.auto_sharding = False
+
+        super().__init__(*args, **kwargs)
+
+        self._connection_state = None
+
+        self._connection_states: list[ConnectionState] = []
+
+        self.max_start_concurrency: int = 1
```

Comment:
It was intentionally not. It could be, but protecting this attribute didnt seem necessary, as the only time you could affect anything with it, would be if you manually login, then start the gateway. That would be an advanced usecase, in which case I'm inclined to allow the modification.